### PR TITLE
8252921: NMT overwrite memory type for region assert when building dynamic archive

### PR DESCRIPTION
**Pull request by zhengyu123**
Thread stack is currently unregistered with NMT in Thread's destructor. Apparently, only Java thread invokes destructor before thread exits. 
For NonJavaThread, e.g. ConcurrentGCThread, thread may exit while its "Thread" object continues alive, therefore, its thread stack is still "alive" from NMT perspective. Once thread exits, the virtual memory for the thread stack can be reserved again, that confused NMT.

The solution is to move thread stack unregistration code to post_run() method.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8252921](https://bugs.openjdk.java.net/browse/JDK-8252921): NMT overwrite memory type for region assert when building dynamic archive

### Reviewers
 * [Yumin Qi](https://openjdk.java.net/census#minqi) (@yminqi - **Reviewer**) ⚠️ Review applies to f4fbe38b70352407271ccb537c772a415ba3ad1b
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/185/head:pull/185`
`$ git checkout pull/185`

### CODE REVIEW

**Review by yminqi — APPROVED** [review]
Looks good.

**Review by calvinccheung — APPROVED** [review]
Thanks for the fix. I've done some testing on the patch. It passed tier1 and also passed running the appcds/dynamicArchive/methodHandles/MethodHandlesAsCollectorTest.java test 30 times on Windows and linux.
One question : why the JavaThread::post_run doesn't need to set the following?  
   set_stack_base(NULL);
   set_stack_size(0);

**Comment by zhengyu123** [comment]
> Thanks for the fix. I've done some testing on the patch. It passed tier1 and also passed running the appcds/dynamicArchive/methodHandles/MethodHandlesAsCollectorTest.java test 30 times on Windows and linux.
> One question : why the JavaThread::post_run doesn't need to set the following?
> set_stack_base(NULL);
> set_stack_size(0);

Thanks for reviewing, Calvin.

The last statement of JavaThread::post_run() deletes 'thread' object, so there is no point to reset its states here. While NonJavaThread object lives pass lifespan of actual thread, its state does matter, I believe.

File: src/hotspot/share/runtime/thread.hpp [hunk 1]
```diff
@@ -759,13 +759,14 @@ class Thread: public ThreadShadow {
 
  public:
   // Stack overflow support
-  address stack_base() const           { assert(_stack_base != NULL,"Sanity check"); return _stack_base; }
+  address stack_base() const           { return _stack_base; }
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.hpp:762]: Why did you remove the assertion? We want the assertion in general to ensure there are no improper uses of stack_base().

Comment (zhengyu123) [reply src/hotspot/share/runtime/thread.hpp:762]: We now reset NonJavaThread's stack base and size,  so _stack_base == NULL is possible, e.g. when generating hs_err report, as we should now see empty stack for "G1 Main Marker" thread in original bug report.

You do have a valid point for the assertion. I think following assertion is more precise:

assert(_stack_base != NULL || Thread::current() != NULL, "Sanity check");

What you think?

Thanks

**Comment by iklam** [comment]
This code in [virtualMemoryTracker.cpp](https://github.com/openjdk/jdk/blob/9a7dcdcdbad34e061a8988287fe691abfd4df305/src/hotspot/share/services/virtualMemoryTracker.cpp#L350) still looks buggy to me:
```
    if (reserved_rgn->same_region(base_addr, size)) {
      reserved_rgn->set_call_stack(stack);
      reserved_rgn->set_flag(flag);
      return true;
    } else {
      assert(reserved_rgn->overlap_region(base_addr, size), "Must be");

      // Overlapped reservation.
      // It can happen when the regions are thread stacks, as JNI
      // thread does not detach from VM before exits, and leads to
      // leak JavaThread object
      if (reserved_rgn->flag() == mtThreadStack) {
        guarantee(!CheckJNICalls, "Attached JNI thread exited without being detached");
        // Overwrite with new region
```
Why is the "Overlapped reservation" fix not done when the two regions are exactly the same? 

If a JNI thread has exited without detaching, and its stack happens to be picked for a future memory reservation of the exact size, I think we will get the same assert as in this bug report.

**Comment by zhengyu123** [comment]
> This code in [virtualMemoryTracker.cpp](https://github.com/openjdk/jdk/blob/9a7dcdcdbad34e061a8988287fe691abfd4df305/src/hotspot/share/services/virtualMemoryTracker.cpp#L350) still looks buggy to me:
> 
> ```
>     if (reserved_rgn->same_region(base_addr, size)) {
>       reserved_rgn->set_call_stack(stack);
>       reserved_rgn->set_flag(flag);
>       return true;
>     } else {
>       assert(reserved_rgn->overlap_region(base_addr, size), "Must be");
> 
>       // Overlapped reservation.
>       // It can happen when the regions are thread stacks, as JNI
>       // thread does not detach from VM before exits, and leads to
>       // leak JavaThread object
>       if (reserved_rgn->flag() == mtThreadStack) {
>         guarantee(!CheckJNICalls, "Attached JNI thread exited without being detached");
>         // Overwrite with new region
> ```
> 
> Why is the "Overlapped reservation" fix not done when the two regions are exactly the same?
> 
The same region branch was to deal with different scenario: recursive reservation, where os::reserve() -> pd_reserve() -> os::reserve()

> If a JNI thread has exited without detaching, and its stack happens to be picked for a future memory reservation of the exact size, I think we will get the same assert as in this bug report.
Right. CheckJNICalls should catch that.

**Comment by iklam** [comment]
> > This code in [virtualMemoryTracker.cpp](https://github.com/openjdk/jdk/blob/9a7dcdcdbad34e061a8988287fe691abfd4df305/src/hotspot/share/services/virtualMemoryTracker.cpp#L350) still looks buggy to me:
> > ```
> >     if (reserved_rgn->same_region(base_addr, size)) {
> >       reserved_rgn->set_call_stack(stack);
> >       reserved_rgn->set_flag(flag);
> >       return true;
> >     } else {
> >       assert(reserved_rgn->overlap_region(base_addr, size), "Must be");
> > 
> >       // Overlapped reservation.
> >       // It can happen when the regions are thread stacks, as JNI
> >       // thread does not detach from VM before exits, and leads to
> >       // leak JavaThread object
> >       if (reserved_rgn->flag() == mtThreadStack) {
> >         guarantee(!CheckJNICalls, "Attached JNI thread exited without being detached");
> >         // Overwrite with new region
> > ```
> > 
> > 
> > Why is the "Overlapped reservation" fix not done when the two regions are exactly the same?
> 
> The same region branch was to deal with different scenario: recursive reservation, where os::reserve() -> pd_reserve() -> os::reserve()
> 
> > If a JNI thread has exited without detaching, and its stack happens to be picked for a future memory reservation of the exact size, I think we will get the same assert as in this bug report.
> 
> Right, there is possibility. We can put the same guarantee there.

I think the order of the checks should be changed:

```
if (reserved_rgn->overlap_region(base_addr, size)) {
    check for overlapped reservation
}
if (reserved_rgn->same_region(base_addr, size)) {
   check for recursive reservation
```
In fact, if the code had been written this way, this bug would not have happened.

**Comment by zhengyu123** [comment]
> > > This code in [virtualMemoryTracker.cpp](https://github.com/openjdk/jdk/blob/9a7dcdcdbad34e061a8988287fe691abfd4df305/src/hotspot/share/services/virtualMemoryTracker.cpp#L350) still looks buggy to me:
> > > ```
> > >     if (reserved_rgn->same_region(base_addr, size)) {
> > >       reserved_rgn->set_call_stack(stack);
> > >       reserved_rgn->set_flag(flag);
> > >       return true;
> > >     } else {
> > >       assert(reserved_rgn->overlap_region(base_addr, size), "Must be");
> > > 
> > >       // Overlapped reservation.
> > >       // It can happen when the regions are thread stacks, as JNI
> > >       // thread does not detach from VM before exits, and leads to
> > >       // leak JavaThread object
> > >       if (reserved_rgn->flag() == mtThreadStack) {
> > >         guarantee(!CheckJNICalls, "Attached JNI thread exited without being detached");
> > >         // Overwrite with new region
> > > ```
> > > 
> > > 
> > > Why is the "Overlapped reservation" fix not done when the two regions are exactly the same?
> > 
> > 
> > The same region branch was to deal with different scenario: recursive reservation, where os::reserve() -> pd_reserve() -> os::reserve()
> > > If a JNI thread has exited without detaching, and its stack happens to be picked for a future memory reservation of the exact size, I think we will get the same assert as in this bug report.
> > 
> > 
> > Right, there is possibility. We can put the same guarantee there.
> 
> I think the order of the checks should be changed:
> 
> ```
> if (reserved_rgn->overlap_region(base_addr, size)) {
>     check for overlapped reservation
> }
> if (reserved_rgn->same_region(base_addr, size)) {
>    check for recursive reservation
> ```
> 
> In fact, if the code had been written this way, this bug would not have happened.

Well, I don't know which bug you refers to, but overlap_region() covers same_region(), so reverse order is a no go.

**Comment by zhengyu123** [comment]
> _Mailing list message from [David Holmes](mailto:david.holmes@oracle.com) on [hotspot-runtime-dev](mailto:hotspot-runtime-dev@openjdk.java.net):_
> 
> Hi Zhengyu,
> 
> On 16/09/2020 9:28 pm, Zhengyu Gu wrote:
> 
> > On Wed, 16 Sep 2020 02:49:45 GMT, David Holmes <dholmes at openjdk.org> wrote:
> > > > Thread stack is currently unregistered with NMT in Thread's destructor. Apparently, only Java thread invokes destructor
> > > > before thread exits.  For NonJavaThread, e.g. ConcurrentGCThread, thread may exit while its "Thread" object continues
> > > > alive, therefore, its thread stack is still "alive" from NMT perspective. Once thread exits, the virtual memory for the
> > > > thread stack can be reserved again, that confused NMT.  The solution is to move thread stack unregistration code to
> > > > post_run() method.
> > > 
> > > 
> > > src/hotspot/share/runtime/thread.hpp line 762:
> > > > 760:  public:
> > > > 761:   // Stack overflow support
> > > > 762:   address stack_base() const           { return _stack_base; }
> > > 
> > > 
> > > Why did you remove the assertion? We want the assertion in general to ensure there are no improper uses of stack_base().
> > 
> > 
> > We now reset NonJavaThread's stack base and size,  so _stack_base == NULL is possible, e.g. when generating hs_err
> > report, as we should now see empty stack for "G1 Main Marker" thread in original bug report.
> 
> Resetting the stack base and size seems pointless given we follow that
> immediately with
> 
> Thread::clear_thread_current();
> 
> And once there is no thread-current for this thread its base and size
> can never be queried, so I don't think this change, or the change to the
> assertion in stack_base() is needed.
> 
> > You do have a valid point for the assertion. I think following assertion is more precise:
> > assert(_stack_base != NULL || Thread::current() != NULL, "Sanity check");
> > What you think?
> 
> That assertion doesn't make sense to me as the stack_base() being
> queried doesn't have to belong to the current thread.

In regular case, yes. But there is an exception, during error reporting (Threads::print_on_error()).
In the original bug report, "G1 Main Marker" thread already exited and should not have stack, but hs_err file shows:

0x000000a47fdee080 ConcurrentGCThread "G1 Main Marker" [stack: 0x000000a41cf40000,0x000000a41d040000] [id=17068]

Thanks,

-Zhengyu


> 
> Thanks,
> David
> -----

**Comment by iklam** [comment]
> > > > This code in [virtualMemoryTracker.cpp](https://github.com/openjdk/jdk/blob/9a7dcdcdbad34e061a8988287fe691abfd4df305/src/hotspot/share/services/virtualMemoryTracker.cpp#L350) still looks buggy to me:
> > > > ```
> > > >     if (reserved_rgn->same_region(base_addr, size)) {
> > > >       reserved_rgn->set_call_stack(stack);
> > > >       reserved_rgn->set_flag(flag);
> > > >       return true;
> > > >     } else {
> > > >       assert(reserved_rgn->overlap_region(base_addr, size), "Must be");
> > > > 
> > > >       // Overlapped reservation.
> > > >       // It can happen when the regions are thread stacks, as JNI
> > > >       // thread does not detach from VM before exits, and leads to
> > > >       // leak JavaThread object
> > > >       if (reserved_rgn->flag() == mtThreadStack) {
> > > >         guarantee(!CheckJNICalls, "Attached JNI thread exited without being detached");
> > > >         // Overwrite with new region
> > > > ```
> > > > 
> > > > 
> > > > Why is the "Overlapped reservation" fix not done when the two regions are exactly the same?
> > > 
> > > 
> > > The same region branch was to deal with different scenario: recursive reservation, where os::reserve() -> pd_reserve() -> os::reserve()
> > > > If a JNI thread has exited without detaching, and its stack happens to be picked for a future memory reservation of the exact size, I think we will get the same assert as in this bug report.
> > > 
> > > 
> > > Right, there is possibility. We can put the same guarantee there.
> > 
> > 
> > I think the order of the checks should be changed:
> > ```
> > if (reserved_rgn->overlap_region(base_addr, size)) {
> >     check for overlapped reservation
> > }
> > if (reserved_rgn->same_region(base_addr, size)) {
> >    check for recursive reservation
> > ```
> > 
> > 
> > In fact, if the code had been written this way, this bug would not have happened.
> 
> Well, I don't know which bug you refer to, but overlap_region() covers same_region(), so reversing order is a no go.

I am talking the original assert reported by [JDK-8252921](https://bugs.openjdk.java.net/browse/JDK-8252921)

```
assert((flag() == mtNone || flag() == f)) failed: Overwrite memory type for region [0x000000a41cf40000-0x000000a41d040000), 3->23.
```

We get there because we have a old stack region which matches exactly as a newly reserved region. This case should also be handled by this code
```
// Overlapped reservation.
// It can happen when the regions are thread stacks, as JNI
// thread does not detach from VM before exits, and leads to
```  

The current code assumes that if you have an exact match, it must be from a recursive reveration. This is wrong.

**Comment by zhengyu123** [comment]
> > > > > This code in [virtualMemoryTracker.cpp](https://github.com/openjdk/jdk/blob/9a7dcdcdbad34e061a8988287fe691abfd4df305/src/hotspot/share/services/virtualMemoryTracker.cpp#L350) still looks buggy to me:
> > > > > ```
> > > > >     if (reserved_rgn->same_region(base_addr, size)) {
> > > > >       reserved_rgn->set_call_stack(stack);
> > > > >       reserved_rgn->set_flag(flag);
> > > > >       return true;
> > > > >     } else {
> > > > >       assert(reserved_rgn->overlap_region(base_addr, size), "Must be");
> > > > > 
> > > > >       // Overlapped reservation.
> > > > >       // It can happen when the regions are thread stacks, as JNI
> > > > >       // thread does not detach from VM before exits, and leads to
> > > > >       // leak JavaThread object
> > > > >       if (reserved_rgn->flag() == mtThreadStack) {
> > > > >         guarantee(!CheckJNICalls, "Attached JNI thread exited without being detached");
> > > > >         // Overwrite with new region
> > > > > ```
> > > > > 
> > > > > 
> > > > > Why is the "Overlapped reservation" fix not done when the two regions are exactly the same?
> > > > 
> > > > 
> > > > The same region branch was to deal with different scenario: recursive reservation, where os::reserve() -> pd_reserve() -> os::reserve()
> > > > > If a JNI thread has exited without detaching, and its stack happens to be picked for a future memory reservation of the exact size, I think we will get the same assert as in this bug report.
> > > > 
> > > > 
> > > > Right, there is possibility. We can put the same guarantee there.
> > > 
> > > 
> > > I think the order of the checks should be changed:
> > > ```
> > > if (reserved_rgn->overlap_region(base_addr, size)) {
> > >     check for overlapped reservation
> > > }
> > > if (reserved_rgn->same_region(base_addr, size)) {
> > >    check for recursive reservation
> > > ```
> > > 
> > > 
> > > In fact, if the code had been written this way, this bug would not have happened.
> > 
> > 
> > Well, I don't know which bug you refer to, but overlap_region() covers same_region(), so reversing order is a no go.
> 
> I am talking the original assert reported by [JDK-8252921](https://bugs.openjdk.java.net/browse/JDK-8252921)
> 
> ```
> assert((flag() == mtNone || flag() == f)) failed: Overwrite memory type for region [0x000000a41cf40000-0x000000a41d040000), 3->23.
> ```
> 
> We get there because we have a old stack region which matches exactly as a newly reserved region. This case should also be handled by this code
> 
> ```
> // Overlapped reservation.
> // It can happen when the regions are thread stacks, as JNI
> // thread does not detach from VM before exits, and leads to
> ```
> 
> The current code assumes that if you have an exact match, it must be from a recursive reveration. This is wrong.

In that case, it just hides the real problem.  Because the thread stack region can be remapped to any other types and maybe different size.

**Comment by iklam** [comment]
> > > > > > This code in [virtualMemoryTracker.cpp](https://github.com/openjdk/jdk/blob/9a7dcdcdbad34e061a8988287fe691abfd4df305/src/hotspot/share/services/virtualMemoryTracker.cpp#L350) still looks buggy to me:
> > > > > > ```
> > > > > >     if (reserved_rgn->same_region(base_addr, size)) {
> > > > > >       reserved_rgn->set_call_stack(stack);
> > > > > >       reserved_rgn->set_flag(flag);
> > > > > >       return true;
> > > > > >     } else {
> > > > > >       assert(reserved_rgn->overlap_region(base_addr, size), "Must be");
> > > > > > 
> > > > > >       // Overlapped reservation.
> > > > > >       // It can happen when the regions are thread stacks, as JNI
> > > > > >       // thread does not detach from VM before exits, and leads to
> > > > > >       // leak JavaThread object
> > > > > >       if (reserved_rgn->flag() == mtThreadStack) {
> > > > > >         guarantee(!CheckJNICalls, "Attached JNI thread exited without being detached");
> > > > > >         // Overwrite with new region
> > > > > > ```
> > > > > > 
> > > > > > 
> > > > > > Why is the "Overlapped reservation" fix not done when the two regions are exactly the same?
> > > > > 
> > > > > 
> > > > > The same region branch was to deal with different scenario: recursive reservation, where os::reserve() -> pd_reserve() -> os::reserve()
> > > > > > If a JNI thread has exited without detaching, and its stack happens to be picked for a future memory reservation of the exact size, I think we will get the same assert as in this bug report.
> > > > > 
> > > > > 
> > > > > Right, there is possibility. We can put the same guarantee there.
> > > > 
> > > > 
> > > > I think the order of the checks should be changed:
> > > > ```
> > > > if (reserved_rgn->overlap_region(base_addr, size)) {
> > > >     check for overlapped reservation
> > > > }
> > > > if (reserved_rgn->same_region(base_addr, size)) {
> > > >    check for recursive reservation
> > > > ```
> > > > 
> > > > 
> > > > In fact, if the code had been written this way, this bug would not have happened.
> > > 
> > > 
> > > Well, I don't know which bug you refer to, but overlap_region() covers same_region(), so reversing order is a no go.
> > 
> > 
> > I am talking the original assert reported by [JDK-8252921](https://bugs.openjdk.java.net/browse/JDK-8252921)
> > ```
> > assert((flag() == mtNone || flag() == f)) failed: Overwrite memory type for region [0x000000a41cf40000-0x000000a41d040000), 3->23.
> > ```
> > 
> > 
> > We get there because we have a old stack region which matches exactly as a newly reserved region. This case should also be handled by this code
> > ```
> > // Overlapped reservation.
> > // It can happen when the regions are thread stacks, as JNI
> > // thread does not detach from VM before exits, and leads to
> > ```
> > 
> > 
> > The current code assumes that if you have an exact match, it must be from a recursive reveration. This is wrong.
> 
> In that case, it just hides the real problem. Because the thread stack region can be remapped to any other types and maybe different size.

Let me clarify. I think your fix is good, but it only fixes part of the problems found by this bug.

I am requesting that the following should **also** be fixed, so that the behavior is the same whether the new region partially overlaps with an old stack, or is exactly the same size as an old stack.

Here's my proposed additional fix. I think your fix alone will not solve the problem where a JNI-attached thread was not properly detached before the thread exits.

```
  if (reserved_rgn == NULL) {
    VirtualMemorySummary::record_reserved_memory(size, flag);
    return _reserved_regions->add(rgn) != NULL;
  } else {
    if (reserved_rgn->overlap_region(base_addr, size) && reserved_rgn->flag() == mtThreadStack) {
      // Overlapped reservation.
      // It can happen when the regions are thread stacks, as JNI
      // thread does not detach from VM before exits, and leads to
      // leak JavaThread object
      guarantee(!CheckJNICalls, "Attached JNI thread exited without being detached");
      // Overwrite with new region

      // Release old region
      VirtualMemorySummary::record_uncommitted_memory(reserved_rgn->committed_size(), reserved_rgn->flag());
      VirtualMemorySummary::record_released_memory(reserved_rgn->size(), reserved_rgn->flag());

      // Add new region
      VirtualMemorySummary::record_reserved_memory(rgn.size(), flag);

      *reserved_rgn = rgn;
      return true;
    } else if (reserved_rgn->same_region(base_addr, size)) {
      reserved_rgn->set_call_stack(stack);
      reserved_rgn->set_flag(flag);
      return true;
    } else {
      assert(reserved_rgn->overlap_region(base_addr, size), "Must be");

      // CDS mapping region.
```

**Comment by zhengyu123** [comment]
> Okay I see the problem now - sorry for the confusion. The issue is when
> printing the other threads. And that means the issue is in the GC thread
> management code as if the GC worker has terminated it should have
> already removed itself from the set of worker threads. Though as this is
> a crash, and we are not at a safepoint, then there could still be a race
> there. :( But note that if we actually deleted the thread we could still
> crash accessing it at that point. So this just seems to be a rare race
> condition that we are unlikely to hit - and if we did then the secondary
> assertion failure would not be too bad IMO.
> 
> But the modified assertion still makes no sense as Thread::current() is
> not relevant AFAICS. I would prefer to keep the assertion to guard the
> regular case of accidental misuse of a thread before the stack_base()
> has been set.
> 
The reported crash actually is at a safepoint, but it is irrelevant.
The problem is that 'Thread' object can outlive thread itself.

"G1 Main Marker" thread is not a worker, but a ConcurrentGCThread, a member of G1CollectedHeap.
<code>G1ConcurrentMarkThread* _cm_thread;</code>
and during error report
<code>Universe::heap()->gc_threads_do(&print_closure);</code>
For example, G1 always tries to report its internal states:
<pre><code>
void G1CollectedHeap::gc_threads_do(ThreadClosure* tc) const {
  workers()->threads_do(tc);
  tc->do_thread(_cm_thread);   <====
  _cm->threads_do(tc);
  _cr->threads_do(tc);
  tc->do_thread(_service_thread);
  if (G1StringDedup::is_enabled()) {
    G1StringDedup::threads_do(tc);
  }
}
</code></pre>

So it is more likely to hit the secondary assertion than you described.

The modified assertion only suggests that you may query thread states from other thread, the result is the best effort and it keeps the semantics unchanged when queries from 'current' thread.

Thanks.

**Comment by zhengyu123** [comment]
Updated to address JNI detach thread issue. Ran tier1, vmTestbase_nsk_monitoring, vmTestbase_nsk_jdi and vmTestbase_nsk_jdwp with NMT on.

**Review by iklam — APPROVED** [review]
Looks good!

File: src/hotspot/share/runtime/thread.cpp [hunk 2]
```diff
@@ -1345,6 +1336,10 @@ void NonJavaThread::pre_run() {
 void NonJavaThread::post_run() {
   JFR_ONLY(Jfr::on_thread_exit(this);)
   remove_from_the_list();
+  unregister_thread_stack_with_NMT();
+  set_stack_base(NULL);
+  set_stack_size(0);
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.cpp:1341]: Sorry but I just don't see any need for clearing stack_base or stack_size, and then you don't need to mess with the assertion in stack_base().

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

**Review by calvinccheung — COMMENTED** [review]
I've tried this patch. It passed tier1 tests and 50 runs of the dynamicArchive/methodHandles/MethodHandlesAsCollectorTest.java test on Windows with vm options: -XX:MaxRAMPercentage=3 -XX:NativeMemoryTracking=detail

File: src/hotspot/share/runtime/thread.cpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/runtime/thread.cpp:1341]: I agree with David that the calls to `set_stack_base(NULL); set_stack_size(0);` do not seem to be related to this bug. Also, previously only `set_stack_base(NULL);` was called -- on all Threads. Now:

- NonJavaThread: `set_stack_base(NULL); set_stack_size(0);`
- JavaThread: neither operation is done

I am wondering:

- why the behavior is changed
- is zeroing the stack_base/size necessary for NMT code only, or is it necessary for non-NMT operations as well?

Will any test cases fail if you remove line 1340/1341?

**Review by iklam — CHANGES_REQUESTED** [review]

File: src/hotspot/share/runtime/thread.cpp (hunk 2 above)
Comment (calvinccheung) [reply src/hotspot/share/runtime/thread.cpp:1341]: I've tried without clearing stack_base and stack_size and also reverted the change in stack_base() in thread.hpp.
It passed tier1 testing. I'll run more test.

Comment (zhengyu123) [reply src/hotspot/share/runtime/thread.cpp:1341]: > Sorry but I just don't see any need for clearing stack_base or stack_size, and then you don't need to mess with the assertion in stack_base().

Well, without cleaning stack_base and stack_size, the thread appears to be live and running, which contradicts NMT report (and maybe others, e.g. smap, etc.)

My initial thought is to filter out terminated thread in Threads::print_on_error(), but it does not seem to have a way.

Comment (zhengyu123) [reply src/hotspot/share/runtime/thread.cpp:1341]: - JavaThread guarantees to delete 'this' before thread exits.
- NonJavaThread object can outlive thread. Zeroing thread stack/size is a way to indicate liveness of the actual thread. I was chasing a race before I realized the thread actual dead, very confusing to see 'G1 Main Marker' as live thread in hs_err file.

**Review by calvinccheung — COMMENTED** [review]
This patch also passed the tests I ran before.

**Review by iklam — APPROVED** [review]
The latest version looks good to me.

**Review by dholmes-ora — APPROVED** [review]
Thanks for restoring those bits. Dealing with the way non-JavaThreads terminate is a known issue (there may already be an open JBS issue for that) and that is best dealt with directly in that context rather than as part of this change.
Thanks.

**Comment by zhengyu123** [comment]
> Thanks for restoring those bits. Dealing with the way non-JavaThreads terminate is a known issue (there may already be an open JBS issue for that) and that is best dealt with directly in that context rather than as part of this change.
> Thanks.

Good to know. Is there a CR for tracking non-JavaThread termination?

Thanks,  David.

**Pull request closed without merging**